### PR TITLE
Require level 1 topics to have level 2 parents; rake tasks to create new topics

### PR DESCRIPTION
**Pull request by cissyxyu**
## WHAT
We're creating a bunch of new Activity Topics. As part of this rollout, we want to require level 1 topics to have level 2 parents from now on. We also want to modify the Activity Editor front end so that when a Curriculum staff member selects a level 2 or level 2 topic, the editor automatically populates that topic's parent and grandparent topics.

## WHY
Part of a rollout of new topics for Activities to add more granularity and specificity to Activity topics.

## HOW
-Rake task to create new Activity Topics.
-Create a .csv of new topics to add. 
-Add a rule to the Topic class that requires level 1 topics to have level 2 parents.
-Modify the Topic editor front end to display and hide not just a selected topic, but also its parent topic and grandparent topic.

### Screenshots
Screen Shot 2022-09-15 at 5 06 15 PM

### Notion Card Links
[(Please provide links to any relevant Notion card(s) relevant to this PR.](https://www.notion.so/quill/Topic-Levels-4e3319bdfa0b42989f632786f9ca92e2)

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES
Have you deployed to Staging? | Not yet - deploying now!
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | Yes
Back-to-school 2022: Have you checked the [webinar schedule](https://www.notion.so/quill/Back-to-school-webinar-banners-2022-a75a89cfad9f434899ef6be3eb184733) to avoid for downtime/risky deploys? | N/A

### CODE REVIEW

File: services/QuillLMS/app/models/topic.rb [hunk 1]
```diff
@@ -64,6 +68,12 @@ def level_three?
   end
 
   private def valid_parent_structure?
-    level_two? ? level_three_parent? : no_parent?
+    if level_one?
```
Comment (brendanshean) [inline services/QuillLMS/app/models/topic.rb]: I would recommend a couple of guard clauses here:

```
return level_two_parent? if level_one?
return level_three_parent? if level_two?

no_parent?
```

File: services/QuillLMS/spec/factories/topic.rb [hunk 2]
```diff
@@ -8,8 +8,11 @@
     parent_id       nil
 
     after(:build) do |t|
-      if t.level == 2
```
Comment (brendanshean) [inline services/QuillLMS/spec/factories/topic.rb:11]: Some cleanup items, most of which is older code:

1. For better readability, use full variable names.
2. To collapse the disjunction, leverage `find_or_create_by`
3. Limit the use of magic numbers by relying on methods `level_two?` and `level_one?`

```
    after(:build) do |topic|
      topic.parent = Topic.find_or_create_by(level: 3) if topic.level_two?
      topic.parent = Topic.find_or_create_by(level: 2) if topic.level_one?
    end
```

File: services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx [hunk 3]
```diff
@@ -39,12 +39,34 @@ const Topics = ({ activity, createNewTopic, topicOptions, handleTopicsChange, })
   }
 
   function onChangeTopics(topicId) {
-    let newTopicIds = [...activity.topic_ids, topicId]
+    let newTopicIds = [...activity.topic_ids]
```
Comment (brendanshean) [inline services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx]: Maybe remove the spread operator: `let newTopicIds = activity.topic_ids`

File: services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx [hunk 4]
```diff
@@ -39,12 +39,34 @@ const Topics = ({ activity, createNewTopic, topicOptions, handleTopicsChange, })
...
     const topic = topicOptions.find(t => t.id === topicId)
     const existingOption = getSelectedOptionForLevel(topic.level)
```
Comment (brendanshean) [inline services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx]: I think the the variable names are making this hard to parse.

Only looking at the names, I would guess there are 3 different types: 
options: existingOption
nodes:  existingParent, existingGrandparent
topics: parentTopic, grandparentTopic

As I read through the logic, it seems like these are all just topic types (or some topic option types).  What about unifying the naming to reflect this:
existingTopic, existingParentTopic, existingGrandparentTopic, parentTopic, grandparentTopic
or
existingOption, existingParentOption, existingGrandparentOption, parentOption, grandparentOption

**Review by brendanshean — APPROVED** [review]
I've left some refactoring comments, but once addressed, I think this should be good to go.  No need for re-review.

File: services/QuillLMS/lib/tasks/create_topics.rake [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+        level_two_topic = Topic.find_or_create_by(name: row["Level 2"], level: 2)
+        level_two_topic.update(parent_id: level_three_topic.id)
+
+        level_one_topic = Topic.find_or_create_by(name: row["Level 1"], level: 1)
+        level_one_topic.update(parent_id: level_two_topic.id)
+      rescue => e
+        puts "Error on Level One Topic: #{row['Level 1']}: #{e}"
+      end
+    end
```
Comment (anathomical) [inline services/QuillLMS/lib/tasks/create_topics.rake]: Minor syntax note.  Because your `begin` block encompasses the entirety of your `each` block, you can use `rescue` without `begin`:
```
    topics_table.each do |row|
      
      level_three_topic = Topic.find_or_create_by(name: row["Level 3"], level: 3)

      level_two_topic = Topic.find_or_create_by(name: row["Level 2"], level: 2)
      level_two_topic.update(parent_id: level_three_topic.id)

      level_one_topic = Topic.find_or_create_by(name: row["Level 1"], level: 1)
      level_one_topic.update(parent_id: level_two_topic.id)
    rescue => e
      puts "Error on Level One Topic: #{row['Level 1']}: #{e}"
    end
```

**Review by anathomical — APPROVED** [review]
Looks good to me!

File: services/QuillLMS/lib/tasks/create_topics.rake [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+namespace :topics do
+  desc 'Create topics from topics table'
+  task :create => :environment do
+    topics_table = CSV.parse(File.read("lib/data/parent_topics.csv"), headers: true)
+    topics_table.each do |row|
+      begin
+        level_three_topic = Topic.find_or_create_by(name: row["Level 3"], level: 3)
+
+        level_two_topic = Topic.find_or_create_by(name: row["Level 2"], level: 2)
+        level_two_topic.update(parent_id: level_three_topic.id)
```
Comment (happythenewsad) [inline services/QuillLMS/lib/tasks/create_topics.rake]: Since you're rescuing and tracking exceptions, and because you added validation logic on Topic, it makes sense to use `update!` in this block.

File: services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx [hunk 7]
```diff
@@ -39,12 +39,34 @@ const Topics = ({ activity, createNewTopic, topicOptions, handleTopicsChange, })
...
+      if (existingParent) {
+        newTopicIds = newTopicIds.filter(id => id !== existingParent.id)
+        const existingGrandparent = topicOptions.find(t => t.id === existingParent.parent_id)
+        if (existingGrandparent) {
+          newTopicIds = newTopicIds.filter(id => id !== existingGrandparent.id)
+        }
+      }
+    }
+
+    if (!existingOption || existingOption.id !== topic.id) {
```
Comment (happythenewsad) [inline services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx]: The cyclomatic complexity of this function is rather large.

**Review by happythenewsad — CHANGES_REQUESTED** [review]
Getting close. Should be good to go after brendan and my comments

File: services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx (hunk 7 above)
Comment (cissyxyu) [reply services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx]: I made some changes to try to simplify it a little -- let me know what you think!

Comment (brendanshean) [reply services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx]: This is looking much better.  One final refactor, consider:

```
if (x) {
  y
} else {
  y
  z
}
```

Since both branches contain `y` we can pull it out:
```
y

if (x) {
} else {
  z
}
```
Since that first conditional is empty now, we can do the negation:

```
y

if not (x) { z }
```
Which in this example reduces to:

```
 newTopicIds = removeExistingTopicAndParents(newTopicIds, previousTopic)

 if (previousTopic?.id !== newTopic.id) { newTopicIds = addNewTopicAndParents(newTopicIds, newTopic) }
```

Comment (cissyxyu) [reply services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx]: Gotcha. Thank you Brendan!

**Review by happythenewsad — APPROVED** [review]

**Review by emilia-friedberg — APPROVED** [review]
